llm/payload.go

```diff
@@ -0,0 +1,230 @@
+package llm
+
+import (
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/jmorganca/ollama/gpu"
+)
+
+var errPayloadMissing = fmt.Errorf("expected payloads not included in this build of ollama")
+
+// TODO: error check this or initialize it in init()
+// TODO: this should be a determinate location to avoid required files disappearing
+var workDir, _ = os.MkdirTemp("", "ollama")
+
+func Init() error {
+	slog.Info("extracting embedded files", "dir", workDir)
+
+	// also extract the .metal shader on darwin/arm64
+	if runtime.GOOS == "darwin" && runtime.GOARCH == "arm64" {
+		err := extractFiles(workDir, "build/*/*/*/bin/ggml-metal.metal.gz")
+		if err != nil {
+			return fmt.Errorf("extract .metal shader: %v", err)
+		}
+
+		os.Setenv("GGML_METAL_PATH_RESOURCES", filepath.Join(workDir, "metal"))
+	}
+
+	var binGlob string
+	if runtime.GOOS == "windows" {
+		binGlob = "build/windows/*/*/bin/*/*"
+	} else {
+		binGlob = "build/*/*/*/bin/*"
+	}
+
+	// extract server libraries
+	err := extractFiles(workDir, binGlob)
+	if err != nil {
+		return fmt.Errorf("extract binaries: %v", err)
+	}
+
+	// TODO: print available variants
+	// slog.Info(fmt.Sprintf("Dynamic LLM libraries %v", variants))
+	// slog.Debug("Override detection logic by setting OLLAMA_LLM_LIBRARY")
+	var variants []string
+	for v := range available() {
+		variants = append(variants, v)
+	}
+	slog.Info(fmt.Sprintf("Dynamic LLM libraries %v", variants))
+	slog.Debug("Override detection logic by setting OLLAMA_LLM_LIBRARY")
+
+	return nil
+}
+
+// binary names may contain an optional variant separated by '_'
+// For example, "ollama_rocm_v6" and "ollama_rocm_v5" or "ollama_cpu" and "ollama_cpu_avx2"
+// Any library without a variant is the lowest common denominator
+func available() map[string]string {
+	// glob workDir for files that start with ollama_
+	pattern := filepath.Join(workDir, "*")
+
+	files, err := filepath.Glob(pattern)
+	if err != nil {
+		slog.Debug("could not glob", "pattern", pattern, "error", err)
+		return nil
+	}
+
+	servers := make(map[string]string)
+	for _, file := range files {
+		slog.Debug("available: found", "file", file)
+		servers[filepath.Base(file)] = file
+	}
+
+	return servers
+}
+
+// serversForGpu returns a list of compatible servers give the provided GPU
+// info, ordered by performance. assumes Init() has been called
+// TODO: complete this given above
+func serversForGpu(info gpu.GpuInfo) []string {
```

Comment:
fwiw I was thinking it probably makes sense at some point (maybe in this change?) to bake some metadata file into the payloads next to each runner so we have a more robust way to select runners.  (what GPU type it is, what CPU compile flags it had, etc.). There's a lot of people asking for non-AVX GPU support, and for that matter, AVX2 GPU runners might also generate enough performance benefit to justify the bloat.

We can certainly come back to it in a follow up, but just calling that out so we don't over-engineer the selection clunky selection algo.